File: connectors/csharp/src/PlayHouse.Connector/Infrastructure/Buffers/RingBuffer.cs
using System;
using System.Buffers;

namespace PlayHouse.Connector.Infrastructure.Buffers;

/// <summary>
/// Pure ring buffer implementation for TCP receive buffering.
/// </summary>
/// <remarks>
/// ⚠️ WARNING: This class is NOT thread-safe.
/// It must be used from a single thread only, or external synchronization must be provided.
/// In this connector, it is used exclusively from the receive loop thread.
/// </remarks>
public sealed class RingBuffer : IDisposable
{
    private static readonly ArrayPool<byte> Pool = ArrayPool<byte>.Shared;

    private byte[] _buffer;
    private int _head;     // Write position
    private int _tail;     // Read position
    private int _count;    // Current data size
    private int _capacity;
    private bool _isDisposed;

    /// <summary>
    /// Create a new RingBuffer with specified capacity
    /// </summary>
    public RingBuffer(int capacity)
    {
        if (capacity <= 0)
        {
            throw new ArgumentException("Capacity must be positive", nameof(capacity));
        }

        _buffer = Pool.Rent(capacity);
        _capacity = _buffer.Length;
        _head = 0;
        _tail = 0;
        _count = 0;
    }

    /// <summary>
    /// Current data size in buffer
    /// </summary>
    public int Count => _count;

    /// <summary>
    /// Buffer capacity
    /// </summary>
    public int Capacity => _capacity;

    /// <summary>
    /// Available free space
    /// </summary>
    public int FreeSpace => _capacity - _count;

    #region Write Operations (Zero-copy)

    /// <summary>
    /// Get writable span for direct writing (Zero-copy)
    /// Returns the maximum contiguous writable space
    /// </summary>
    public Span<byte> GetWriteSpan(int requestedSize)
    {
        if (requestedSize <= 0)
        {
            throw new ArgumentException("Size must be positive", nameof(requestedSize));
        }

        if (requestedSize > FreeSpace)
        {
            throw new InvalidOperationException($"Not enough free space. Requested: {requestedSize}, Available: {FreeSpace}");
        }

        // Calculate contiguous writable size
        int contiguousSize;
        if (_head >= _tail || _count == 0)
        {
            // Case 1: [___TTTT____HH___] or empty buffer
            // Can write from head to end of buffer
            contiguousSize = _capacity - _head;

            // If wrapping is needed and requested size is larger than tail space
            if (contiguousSize < requestedSize && _tail > 0)
            {
                // Return only the space until end, caller should call again for wrapped part
                contiguousSize = Math.Min(contiguousSize, requestedSize);
            }
            else
            {
                contiguousSize = Math.Min(contiguousSize, requestedSize);
            }
        }
        else
        {
            // Case 2: [HH____TTTT]
            // Can write from head to tail
            contiguousSize = Math.Min(_tail - _head, requestedSize);
        }

        return _buffer.AsSpan(_head, contiguousSize);
    }

    /// <summary>
    /// Advance write position after writing data
    /// </summary>
    public void Advance(int count)
    {
        if (count < 0)
        {
            throw new ArgumentException("Count must be non-negative", nameof(count));
        }

        if (count > FreeSpace)
        {
            throw new InvalidOperationException($"Cannot advance by {count}, only {FreeSpace} free space available");
        }

        _head = (_head + count) % _capacity;
        _count += count;
    }

    /// <summary>
    /// Write single byte
    /// </summary>
    public void WriteByte(byte value)
    {
        if (FreeSpace < 1)
        {
            throw new InvalidOperationException("Buffer is full");
        }

        _buffer[_head] = value;
        _head = (_head + 1) % _capacity;
        _count++;
    }

    /// <summary>
    /// Write bytes to buffer
    /// </summary>
    public void WriteBytes(ReadOnlySpan<byte> data)
    {
        if (data.Length > FreeSpace)
        {
            throw new InvalidOperationException($"Not enough space to write {data.Length} bytes");
        }

        var remaining = data.Length;
        var offset = 0;

        while (remaining > 0)
        {
            var writeSpan = GetWriteSpan(remaining);
            var writeCount = Math.Min(writeSpan.Length, remaining);

            data.Slice(offset, writeCount).CopyTo(writeSpan);
            Advance(writeCount);

            offset += writeCount;
            remaining -= writeCount;
        }
    }

    #endregion

    #region Read Operations (Zero-copy)

    /// <summary>
    /// Peek at data without consuming (Zero-copy)
    /// </summary>
    public ReadOnlySpan<byte> Peek(int count)
    {
        if (count < 0)
        {
            throw new ArgumentException("Count must be non-negative", nameof(count));
        }

        if (count > _count)
        {
            throw new InvalidOperationException($"Cannot peek {count} bytes, only {_count} available");
        }

        if (count == 0)
        {
            return ReadOnlySpan<byte>.Empty;
        }

        // Calculate contiguous readable size
        int contiguousSize;
        if (_tail < _head || (_tail == _head && _count == 0))
        {
            // Case 1: [___TTTT____HH___]
            contiguousSize = Math.Min(_head - _tail, count);
        }
        else
        {
            // Case 2: [HH____TTTT] or buffer is full
            contiguousSize = Math.Min(_capacity - _tail, count);
        }

        return new ReadOnlySpan<byte>(_buffer, _tail, contiguousSize);
    }

    /// <summary>
    /// Peek byte at specific offset from current tail
    /// </summary>
    public byte PeekByte(int offset)
    {
        if (offset < 0 || offset >= _count)
        {
            throw new ArgumentOutOfRangeException(nameof(offset));
        }

        var index = (_tail + offset) % _capacity;
        return _buffer[index];
    }

    /// <summary>
    /// Peek multiple bytes at specific offset (handles wrap-around for bulk copy)
    /// Reduces function call overhead from N PeekByte() calls to 1-2 memory copies
    /// </summary>
    /// <param name="offset">Offset from current tail position</param>
    /// <param name="destination">Destination span to copy bytes into</param>
    public void PeekBytes(int offset, Span<byte> destination)
    {
        if (offset < 0 || offset + destination.Length > _count)
        {
            throw new ArgumentOutOfRangeException(nameof(offset),
                $"Cannot peek {destination.Length} bytes at offset {offset}, only {_count} bytes available");
        }

        if (destination.Length == 0)
        {
            return;
        }

        // Calculate read start position in circular buffer
        var readStart = (_tail + offset) % _capacity;

        // First chunk: from readStart to end of buffer or end of destination
        var firstChunkSize = Math.Min(_capacity - readStart, destination.Length);
        _buffer.AsSpan(readStart, firstChunkSize).CopyTo(destination);

        // Second chunk: if data wraps around to beginning of buffer
        if (firstChunkSize < destination.Length)
        {
            var secondChunkSize = destination.Length - firstChunkSize;
            _buffer.AsSpan(0, secondChunkSize).CopyTo(destination.Slice(firstChunkSize));
        }
    }

    /// <summary>
    /// Consume (remove) data from buffer after reading
    /// </summary>
    public void Consume(int count)
    {
        if (count < 0)
        {
            throw new ArgumentException("Count must be non-negative", nameof(count));
        }

        if (count > _count)
        {
            throw new InvalidOperationException($"Cannot consume {count} bytes, only {_count} available");
        }

        _tail = (_tail + count) % _capacity;
        _count -= count;
    }

    /// <summary>
    /// Read single byte
    /// </summary>
    public byte ReadByte()
    {
        if (_count == 0)
        {
            throw new InvalidOperationException("Buffer is empty");
        }

        var value = _buffer[_tail];
        _tail = (_tail + 1) % _capacity;
        _count--;
        return value;
    }

    /// <summary>
    /// Read bytes into destination buffer
    /// </summary>
    public int ReadBytes(Span<byte> destination)
    {
        var toRead = Math.Min(destination.Length, _count);
        var remaining = toRead;
        var offset = 0;

        while (remaining > 0)
        {
            var peekSpan = Peek(remaining);
            var readCount = Math.Min(peekSpan.Length, remaining);

            peekSpan.Slice(0, readCount).CopyTo(destination.Slice(offset));
            Consume(readCount);

            offset += readCount;
            remaining -= readCount;
        }

        return toRead;
    }

    #endregion

    #region Utility Methods

    /// <summary>
    /// Clear all data in buffer
    /// </summary>
    public void Clear()
    {
        _head = 0;
        _tail = 0;
        _count = 0;
    }

    /// <summary>
    /// Get next index in circular buffer
    /// </summary>
    private int NextIndex(int index)
    {
        return (index + 1) % _capacity;
    }

    #endregion

    public void Dispose()
    {
        if (_isDisposed)
        {
            return;
        }

        if (_buffer != null)
        {
            Pool.Return(_buffer);
            _buffer = null!;
        }

        _isDisposed = true;
    }
}

File: connectors/cpp/src/ring_buffer.cpp
#include "ring_buffer.hpp"
#include <algorithm>
#include <cstring>
#include <stdexcept>

namespace playhouse {
namespace internal {

RingBuffer::RingBuffer(size_t capacity)
    : buffer_(capacity)
    , capacity_(capacity)
    , head_(0)
    , tail_(0)
    , count_(0)
{}

size_t RingBuffer::GetCount() const {
    return count_;
}

size_t RingBuffer::GetCapacity() const {
    return capacity_;
}

size_t RingBuffer::GetFreeSpace() const {
    return capacity_ - count_;
}

void RingBuffer::Write(const uint8_t* data, size_t size) {
    if (size > GetFreeSpace()) {
        throw std::runtime_error("Not enough space in ring buffer");
    }

    if (size == 0) {
        return;
    }

    // Calculate how much we can write contiguously
    size_t contiguous_space = capacity_ - head_;
    size_t first_chunk = std::min(size, contiguous_space);

    // Write first chunk
    std::memcpy(buffer_.data() + head_, data, first_chunk);
    head_ = (head_ + first_chunk) % capacity_;
    count_ += first_chunk;

    // Write second chunk if needed (wrap around)
    if (first_chunk < size) {
        size_t second_chunk = size - first_chunk;
        std::memcpy(buffer_.data() + head_, data + first_chunk, second_chunk);
        head_ = (head_ + second_chunk) % capacity_;
        count_ += second_chunk;
    }
}

void RingBuffer::Read(uint8_t* dest, size_t size) {
    if (size > count_) {
        throw std::runtime_error("Not enough data in ring buffer");
    }

    if (size == 0) {
        return;
    }

    // Calculate how much we can read contiguously
    size_t contiguous_data = capacity_ - tail_;
    size_t first_chunk = std::min(size, contiguous_data);

    // Read first chunk
    std::memcpy(dest, buffer_.data() + tail_, first_chunk);
    tail_ = (tail_ + first_chunk) % capacity_;
    count_ -= first_chunk;

    // Read second chunk if needed (wrap around)
    if (first_chunk < size) {
        size_t second_chunk = size - first_chunk;
        std::memcpy(dest + first_chunk, buffer_.data() + tail_, second_chunk);
        tail_ = (tail_ + second_chunk) % capacity_;
        count_ -= second_chunk;
    }
}

void RingBuffer::Peek(uint8_t* dest, size_t size, size_t offset) const {
    if (offset + size > count_) {
        throw std::runtime_error("Not enough data in ring buffer to peek");
    }

    if (size == 0) {
        return;
    }

    // Calculate read position
    size_t read_pos = (tail_ + offset) % capacity_;

    // Calculate how much we can read contiguously
    size_t contiguous_data = capacity_ - read_pos;
    size_t first_chunk = std::min(size, contiguous_data);

    // Read first chunk
    std::memcpy(dest, buffer_.data() + read_pos, first_chunk);

    // Read second chunk if needed (wrap around)
    if (first_chunk < size) {
        size_t second_chunk = size - first_chunk;
        std::memcpy(dest + first_chunk, buffer_.data(), second_chunk);
    }
}

uint8_t RingBuffer::PeekByte(size_t offset) const {
    if (offset >= count_) {
        throw std::runtime_error("Offset out of range");
    }

    size_t pos = (tail_ + offset) % capacity_;
    return buffer_[pos];
}

void RingBuffer::Consume(size_t size) {
    if (size > count_) {
        throw std::runtime_error("Cannot consume more data than available");
    }

    tail_ = (tail_ + size) % capacity_;
    count_ -= size;
}

void RingBuffer::Clear() {
    head_ = 0;
    tail_ = 0;
    count_ = 0;
}

const uint8_t* RingBuffer::GetReadPtr() const {
    return buffer_.data() + tail_;
}

size_t RingBuffer::GetContiguousReadSize() const {
    if (count_ == 0) {
        return 0;
    }

    // When buffer is full (count_ == capacity_), head_ == tail_
    // In this case, data wraps around and we return from tail to end
    if (count_ == capacity_) {
        return capacity_ - tail_;
    }

    if (tail_ < head_) {
        // Data is contiguous
        return head_ - tail_;
    } else {
        // Data wraps around
        return capacity_ - tail_;
    }
}

uint8_t* RingBuffer::GetWritePtr() {
    return buffer_.data() + head_;
}

size_t RingBuffer::GetContiguousWriteSize() const {
    size_t free_space = GetFreeSpace();
    if (free_space == 0) {
        return 0;
    }

    if (head_ >= tail_ || count_ == 0) {
        // Can write from head to end or to tail
        size_t to_end = capacity_ - head_;
        if (count_ == 0) {
            return to_end;
        }
        return std::min(to_end, free_space);
    } else {
        // Can write from head to tail
        return tail_ - head_;
    }
}

void RingBuffer::Advance(size_t size) {
    if (size > GetFreeSpace()) {
        throw std::runtime_error("Cannot advance more than free space");
    }

    head_ = (head_ + size) % capacity_;
    count_ += size;
}

} // namespace internal
} // namespace playhouse
